stdlib/public/core/StringComparable.swift

```diff
@@ -178,6 +207,31 @@ extension _StringGuts {
     return _compareDeterministicUnicodeCollation(
       left, leftRange,
       to: right, rightRange)
+#endif
+  }  
+
+  @_inlineable
+  @_versioned
+  internal static func compare(
+    _ left: _StringGuts, to right: _StringGuts
+  ) -> Int {
+#if _runtime(_ObjC)
+    // We only want to perform this optimization on objc runtimes. Elsewhere,
+    // we will make it follow the unicode collation algorithm even for ASCII.
+    // This is consistent with Foundation, but incorrect as defined by Unicode.
+    //
+    // FIXME: String ordering should be consistent across all platforms.
+    if left.isASCII && right.isASCII {
+      let leftASCII = left._unmanagedASCIIView
+      let rightASCII = right._unmanagedASCIIView
+      let result = leftASCII.compareASCII(to: rightASCII)
+      _fixLifetime(left)
+      _fixLifetime(right)
+      return result
+    }
+    return _compareDeterministicUnicodeCollation(left, to: right)
+#else
+    return _compareDeterministicUnicodeCollation(left, to: right)
```

Comment:
@lorentey, I'm not a big fan of duplicating this like this :-(. But, the ranges were showing a ~40% overhead on some benchmarks, even before I fixed the excessive ARC in them. Any ideas on how to better reuse the code?

Reply:
I think we should define these operations on the unmanaged views; they are self-sliceable, so that takes care of  partial compares, but adds a combinatorial explosion in the number of cases.

I ran into the same problem while updating grapheme braking, and I'm experimenting with adding an internal _StringVariant protocol abstracting the three unmanaged views into a single interface; I think we will be able to define a single `compare` function as an extension on that.  (All three unmanaged views are RandomAccessCollections of UInt16 values; they all support zero-based indexing and extracting a self-slice over a range of integer offsets, etc.)